# 📚 Random Forest with Out-of-Bag (OOB) Score

### 🚀 Goal:
- Train a **Random Forest Classifier** and check its **OOB (Out-of-Bag) score** for quick model evaluation.

---

### 📦 Breakdown:

```python
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification
```
- **Import** the classifier and a dataset generator.

```python
X, y = make_classification(n_samples=1000, n_features=20, n_classes=2, random_state=42)
```
- **Create dummy data** with 1000 samples, 20 features.

```python
rf_classifier = RandomForestClassifier(n_estimators=100, oob_score=True, random_state=42)
```
- **Random Forest** with:
  - 100 trees,
  - **OOB scoring enabled** (evaluates performance using unused samples in bootstrapping),
  - Random state for reproducibility.

```python
rf_classifier.fit(X, y)
```
- **Train** the model.

```python
oob_score = rf_classifier.oob_score_
print("Out-of-Bag Score:", oob_score)
```
- **Print OOB Score** (an internal validation score, like cross-validation).

---

### 🧠 Quick Tip:
**OOB Score = Free cross-validation using leftover (out-of-bag) samples!**

In [1]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification


X, y = make_classification(n_samples=1000, n_features=20, n_classes=2, random_state=42)


rf_classifier = RandomForestClassifier(n_estimators=100, oob_score=True, random_state=42)


rf_classifier.fit(X, y)

oob_score = rf_classifier.oob_score_
print("Out-of-Bag Score:", oob_score)


Out-of-Bag Score: 0.895


# 📚 Out-of-Bag (OOB) Score Insights....

- **OOB Score** is the model’s accuracy using unused samples during training.
- **0.895 OOB** means ~89.5% correct predictions without needing a test set.
- It acts like a **free cross-validation** inside Random Forest.
- Useful for **quick model evaluation**.